core/types/transaction_signing.go

```diff
@@ -0,0 +1,273 @@
+	Hash(tx *Transaction) common.Hash
+	// PubilcKey returns the public key derived from the signature
+	PublicKey(tx *Transaction) ([]byte, error)
+	// SignECDSA signs the transaction with the given and returns a copy of the tx
+	SignECDSA(tx *Transaction, prv *ecdsa.PrivateKey) (*Transaction, error)
+	// WithSignature returns a copy of the transaction with the given signature
+	WithSignature(tx *Transaction, sig []byte) (*Transaction, error)
+}
+
+// EIP155Transaction implements TransactionInterface using the
+// EIP155 rules
+type ChainIdSigner struct {
+	BasicSigner
+
+	chainId, chainIdMul *big.Int
+}
+
+func NewChainIdSigner(chainId *big.Int) ChainIdSigner {
+	return ChainIdSigner{
+		chainId:    chainId,
+		chainIdMul: new(big.Int).Mul(chainId, big.NewInt(2)),
+	}
+}
+
+func (s ChainIdSigner) SignECDSA(tx *Transaction, prv *ecdsa.PrivateKey) (*Transaction, error) {
+	return SignECDSA(s, tx, prv)
+}
+
+func (s ChainIdSigner) PublicKey(tx *Transaction) ([]byte, error) {
+	// if the transaction is not protected fall back to homestead signer
+	if !tx.Protected() {
+		return (BasicSigner{}).PublicKey(tx)
+	}
+	if tx.ChainId().Cmp(s.chainId) != 0 {
+		return nil, ErrInvalidChainId
+	}
+
+	V := normaliseV(s, tx.data.V)
+	if !crypto.ValidateSignatureValues(V, tx.data.R, tx.data.S, true) {
+		return nil, ErrInvalidSig
+	}
+
+	// encode the signature in uncompressed format
+	R, S := tx.data.R.Bytes(), tx.data.S.Bytes()
+	sig := make([]byte, 65)
+	copy(sig[32-len(R):32], R)
+	copy(sig[64-len(S):64], S)
+	sig[64] = V - 27
+
+	// recover the public key from the signature
+	hash := s.Hash(tx)
+	pub, err := crypto.Ecrecover(hash[:], sig)
+	if err != nil {
+		return nil, err
+	}
+	if len(pub) == 0 || pub[0] != 4 {
+		return nil, errors.New("invalid public key")
+	}
+	return pub, nil
+}
+
+// WithSignature returns a new transaction with the given signature.
+// This signature needs to be formatted as described in the yellow paper (v+27).
+func (s ChainIdSigner) WithSignature(tx *Transaction, sig []byte) (*Transaction, error) {
+	if len(sig) != 65 {
+		panic(fmt.Sprintf("wrong size for snature: got %d, want 65", len(sig)))
+	}
+
+	cpy := &Transaction{signer: tx.signer, data: tx.data}
+	cpy.data.R = new(big.Int).SetBytes(sig[:32])
+	cpy.data.S = new(big.Int).SetBytes(sig[32:64])
+	cpy.data.V = new(big.Int).SetBytes([]byte{sig[64]})
+	if s.chainId.BitLen() > 0 {
+		cpy.data.V = big.NewInt(int64(sig[64] + 35))
+		cpy.data.V.Add(cpy.data.V, s.chainIdMul)
+	}
+	return cpy, nil
+}
+
+// Hash returns the hash to be signed by the sender.
+// It does not uniquely identify the transaction.
+func (s ChainIdSigner) Hash(tx *Transaction) common.Hash {
+	return rlpHash([]interface{}{
+		tx.data.AccountNonce,
+		tx.data.Price,
+		tx.data.GasLimit,
+		tx.data.Recipient,
+		tx.data.Amount,
+		tx.data.Payload,
+		s.chainId, uint(0), uint(0),
+	})
+}
+
+func (s ChainIdSigner) SigECDSA(tx *Transaction, prv *ecdsa.PrivateKey) (*Transaction, error) {
```

Comment:
Ah, you mean that it's same just call another function, so looks different from BasicSign? Yeah, it seems that it requires some refactoring